src/rest/connection.cpp

```diff
@@ -0,0 +1,357 @@
+/*
+ *  Copyright (c) 2020, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "rest/connection.hpp"
+
+#include <cerrno>
+
+#include <assert.h>
+#include <sys/time.h>
+
+using std::chrono::duration_cast;
+using std::chrono::microseconds;
+using std::chrono::seconds;
+using std::chrono::steady_clock;
+
+namespace otbr {
+namespace rest {
+
+// The timeout (in microseconds) since a connection is in wait callback state
+static const uint32_t kCallbackTimeout = 10000000;
+// The time interval (in microseconds) for checking again if there is a connection need callback.
+static const uint32_t kCallbackCheckInterval = 500000;
+// The timeout (in microseconds) since a connection is in wait write state
+static const uint32_t kWriteTimeout = 10000000;
+// The timeout (in microseconds) since a connection is in wait read state
+static const uint32_t kReadTimeout = 1000000;
+
+Connection::Connection(steady_clock::time_point aStartTime, Resource *aResource, int aFd)
+    : mStartTime(aStartTime)
+    , mFd(aFd)
+    , mState(OTBR_REST_CONNECTION_INIT)
```

Comment:
So for example, does ConnectionState naming should start with  `OTBR_REST` like `OTBR_REST_INIT`

Reply:
I think the intention is clear. No need for prefix like `OTBR_TEST`, because the C++ namespace (otbr::Rest) already serves this purpose. Prefixes are only required for macros, because C++ namespace doesn't apply to macros.